src/parquet/file/reader-internal.cc

```diff
@@ -174,13 +175,19 @@ std::unique_ptr<PageReader> SerializedRowGroup::GetColumnPageReader(int i) {
   }
 
   int64_t bytes_to_read = col.meta_data.total_compressed_size;
-  std::shared_ptr<Buffer> buffer = source_->ReadAt(col_start, bytes_to_read);
+  std::unique_ptr<InputStream> stream;
 
-  if (buffer->size() < bytes_to_read) {
-    throw ParquetException("Unable to read column chunk data");
+  if (chunk_size == 0) {
+    std::shared_ptr<Buffer> buffer = source_->ReadAt(col_start, bytes_to_read);
+    if (buffer->size() < bytes_to_read) {
+      throw ParquetException("Unable to read column chunk data");
+    }
+    stream.reset(new InMemoryInputStream(buffer));
+  } else {
+    stream.reset(new ChunkedInMemoryInputStream(source_, allocator_, col_start,
+        col_start + bytes_to_read, chunk_size));
```

Comment:
Per the PR-level discussion about simplifying the RandomAccessSource API, we could add a `source_->ReadChunkedStream(...)` API along with `ReadStream`. I'm not sure where is the best place to put the logic to switch between chunked and non-chunked streams (it would be better to have it in *exactly one* place rather than each place you might like to have both options)